### ISSUE TITLE
RFC : how to support multi-platform message channel ?

### ISSUE THREAD

**Issue by wj-Mcat** [wechaty/wechaty#2023]
There days, we get deep into the multi-platform `group chat bot` documents and find that there are some message types and message payload datastructure which are different from the current message type in `wechaty-puppet`. In other words, the current `wechaty-puppet` can support the `rich` `message info` in `dingtalk`, `feishu`, `enterprise wechat`. 

I have list some example message payloads: 
- [feishu message data structure](http://note.youdao.com/s/CNbReEVG) 
- [dingtalk message data structure](http://note.youdao.com/s/BFFs7lcO) 
- [enterprise message data structure](http://note.youdao.com/s/DjqDSPlL) 

So, I think we should add some message types and message payload to support more multi-platform wechaty. How do you think about it ?

**Comment by windmemory** [comment]
I would prefer to have several message types that are currently supported by all these three platforms, as `Wechaty` is targeting a universe chatbot connector, the types that are mostly in common are those types that we want to do first.

From my perspective, I think the `AcitonCard`(card with option buttons) and `FeedCard`(card with multiple links inside) are those types that we should go first.

@huan what do you think about these new message types?

**Comment by huan** [comment]
@wj-Mcat Thank you very much for create this RFC for discussing!

I agree that when we start supporting multi-platform puppet protocols other than Wechaty (like Feishu, DingTalk, etc), the payload might need to be reviewed and we need to find a way for compatible with multi protocols.

@windmemory I fully agree with you that our Wechaqty targeting a universe chatbot connector, and we should have a universe payload for compatible with all IMs, as ActionCard is also needed by the Microsoft BotFramework and the Facebook Messenger.

Please feel free to submit a solution by designing a new payload system for our UrlLink payload (or new payload if necessary), then we will be able to discuss those payloads based on our first version.

P.S. @wj-Mcat I'd like to suggest that we should always post all the requires materials in the issue so that we can read them in one place. for example, your example message payloads can be easy pasted into our issue markdown.

**Comment by wj-Mcat** [comment]
@huan Yes, I will post all of the mutli-platform detailed message payloads example code at public place. In this discuss, ***how to support multi-platform message channel ?*** is a multi-steps for wechaty community.  Here is the idea discussed with @windmemory : 
1. **First**, we should try to run the robot of each platform and  look at the data structure inside and analyze the details of each platform.
2. **Second**, let's take a look at the similarities and differences among the platforms, discuss them, and develop a generic, larger design for wechaty.
3. **Third**, we can then develop it on one of the wechaty language platform, eg: [`ts-wechaty`](https://github.com/wechaty/wechaty). Maybe we should modify some of puppet's abstraction layer design, but I thinks this is normal. Once it becomes stable, other multilingual Wechaty can be translated directly.

@huan How do you think about it ?

**Comment by huan** [comment]
I definitely agree with you guys about all those points.

Please go ahead and let me know if you need my help!
